### ISSUE TITLE
False positive: Avoid unused private method when using it as a callback in a function

### ISSUE THREAD

**Issue by eleracmp** [phpmd/phpmd#897]
- PHPMD version: 2.9.1
- PHP Version: 7.4
- Installation type: (composer)
- Operating System / Distribution & Version: (Ubuntu)

### Current Behavior
```php
return array_map([$this, 'myPrivateMethod'], $array);
```
I have an error on the private method myPrivateMethod who is avoid unused private method. Phpmd don't detect that [$this, 'myPrivateMethod'] is referencing that method

### Expected Behavior
phpmd detect that the method is in use and don't make the false positive. 

### Steps To Reproduce:
Create a class with a private method. use that method as a callback somewhere in the same class by the notation [$this,'method']. For example like return array_map([$this, 'myPrivateMethod'], $array);

**Comment by kylekatarnls** [comment]
Hello, 👋 

We have a similar open thread: #502

Basically, by reading only the expression `[$this, 'myPrivateMethod']` this is not possible to conclude that this array represent a potential call to `myPrivateMethod()` of the current class. So we have to involve the typing of the function (e.g. we have to read that `array_map` expect a `callable` as a parameter). Then as it can be stored in a variable:
```php
$func = [$this, 'myPrivateMethod'];
array_map($func, $array);
```
Then in theory, we should also detect this use.

Last `[$this, 'myPrivateMethod']` can have the first or the second argument as the result of an expression.

Plus the same way, we should handle the string syntax which is also valid (as per #501):
```php
array_map('MyClass::myPrivateMethod', $array);
```

Considering we can only partially fix it (only the static value + explicit `$this`) and we have a high cost for this change (tracking `callable` parameters of functions/methods), we did not invest time in this.

Meanwhile we recommend to use explicit calls:
```php
array_map(fn ($value) => $this->myPrivateMethod($value), $array);
```

This is actually a better option regarding best practices and to avoid future bugs in your app.

See the following situation:
```php
public function getRand(): int
{
  return mt_rand(0, 9);
}

public function foo(array $array): array
{
  // Replace an array with an array of the same length with random values
  return array_map(fn () => $this->getRand(), $array); // arrow function option
  return array_map([$this, 'getRand'], $array); // array callback option
}
```
You get the expected behavior, get your array with N random digits

Now for the need of an other function somewhere else, you (or an other colleague) need to implement a new parameter for `getRand()`:

```php
public function getRand(int $max = 9): int
{
  return mt_rand(0, $max);
}
```
This change is not supposed to be a breaking change, adding parameters to a method should be safe as one does not expect a parameter that was not needed to be passed to the function, also if you try to search in your code base for existing calls of getRand with a param to be sure with a RegExp search like `getRand\([^)]` you will find any `getRand(5` or `getRand($foo` but you won't get `'getRand'` in your search result.

If the change above is pushed, the a method using `array_map(fn () => $this->getRand(), $array)` is safe, it will still return the same result as before.

But `array_map([$this, 'getRand'], $array)` will now use the passed value as the max of getRand, it's it's number you'll get an incorrect maximum for your random number, if it's an other type, you'll simply get a type error exception.

That's why in my projects, I avoid to use the `[$this, 'getRand']` which is risky when parameters evolve and also is less discoverable (IDE search, static analyze, ease for new developers to get it).

Of course this does not mean we should not support the syntax, but it actually make this issue a low priority from our perspective. Still we'll help reviewing the pull-request if someone want to give a try to fixing this.

Thanks.

**Comment by kylekatarnls** [comment]
Fix proposal submitted as #1068
